perl-package/AI-MXNet/lib/AI/MXNet/Gluon/NN/ConvLayers.pm

```diff
@@ -0,0 +1,1363 @@
+has 'op_name'            => (is => 'rw', isa => 'Str', default => 'Convolution');
+has 'use_bias'           => (is => 'rw', isa => 'Bool', default => 1);
+has 'weight_initializer' => (is => 'rw', isa => 'Maybe[Initializer]');
+has 'bias_initializer'   => (is => 'rw', isa => 'Maybe[Initializer]', default => 'zeros');
+has 'adj'                => (is => 'rw');
+has [qw/weight bias
+        kwargs act/]     => (is => 'rw', init_arg => undef);
+method python_constructor_arguments() { [qw/channels kernel_size strides padding dilation/] }
+
+sub BUILD
+{
+    my $self = shift;
+    $self->_update_kernel_size;
+    $self->name_scope(sub {
+        if(not ref $self->strides)
+        {
+            $self->strides([($self->strides) x @{ $self->kernel_size }]);
+        }
+        if(not ref $self->padding)
+        {
+            $self->padding([($self->padding) x @{ $self->kernel_size }]);
+        }
+        if(not ref $self->dilation)
+        {
+            $self->dilation([($self->dilation) x @{ $self->kernel_size }]);
+        }
+        $self->kwargs({
+            kernel => $self->kernel_size, stride => $self->strides, dilate => $self->dilation,
+            pad => $self->padding, num_filter => $self->channels, num_group => $self->groups,
+            no_bias => $self->use_bias ? 0 : 1, layout => $self->layout
+        });
+        if(defined $self->adj)
+        {
+            $self->kwargs->{adj} = $self->adj;
+        }
+
+        my @dshape = (0)x(@{ $self->kernel_size } + 2);
+        $dshape[index($self->layout, 'N')] = 1;
+        $dshape[index($self->layout, 'C')] = $self->in_channels;
+        my $wshapes = _infer_weight_shape($self->op_name, \@dshape, $self->kwargs);
+        $self->weight(
+            $self->params->get(
+                'weight', shape => $wshapes->[1],
+                init => $self->weight_initializer,
+                allow_deferred_init => 1
+            )
+        );
+        if($self->use_bias)
+        {
+            $self->bias(
+                $self->params->get(
+                    'bias', shape => $wshapes->[2],
+                    init => $self->bias_initializer,
+                    allow_deferred_init => 1
+                )
+            );
+        }
+        if(defined $self->activation)
+        {
+            $self->act(
+                AI::MXNet::Gluon::NN::Activation->new(
+                    activation => $self->activation,
+                    prefix     => $self->activation.'_'
+                )
+            );
+        }
+    });
+}
+
+method hybrid_forward(GluonClass $F, GluonInput $x, GluonInput :$weight, Maybe[GluonInput] :$bias=)
+{
+    my $op_name = $self->op_name;
+    my $act = $F->$op_name($x, $weight, defined $bias ? $bias : (), name => 'fwd', %{ $self->kwargs });
+    if(defined $self->act)
+    {
+        $act = $self->act->($act);
+    }
+    return $act;
+}
+
+method _alias() { 'conv' }
+
+use Data::Dumper;
+use overload '""' => sub {
```

Comment:
translating `def __repr__` to `use overload '""' => sub { ... }` may cause some collisions if the python guys add a `def __str__` later on or if the algorithms expects str(...) to be more performant than the `__repr__` implementation.  an explicit object method might make sense, but obviously it would be a hassle to track down all the places we'd need to invoke it where python was calling it implicitly.

Reply:
yes, let's worry about this later.

Reply:
but we need bool overload for these objects that's for sure. I'll add that later, or may be update mouse to have global bool overload aliased to defined.